### ISSUE TITLE
RPC call cannot set headers in 1.0.0-dev.4

### ISSUE THREAD

**Issue by dsyrstad** [supabase-community/supabase-flutter#197]

**Comment by dsyrstad** [comment]
> @dsyrstad It sounds like what you really want is a handy way of setting `Prefer: params=single-object` header.

That would be even better. If you're considering that, maybe you should also add: `Prefer: params=multiple-objects`  https://postgrest.org/en/stable/api.html#bulk-call.

**Comment by grdsdev** [comment]
Hey there, we’re actively working on reducing the number of open issues. Since this issue is quite old, I’ll be closing it. However, feel free to reopen it if you still believe it’s valid.

Thanks.

**Comment by dsyrstad** [comment]
@grdsdev We are still actively working around this issue and would really like to see an official implementation, like @dshukertjr mentioned in https://github.com/supabase/supabase-flutter/issues/197#issuecomment-1221563138.

**Comment by grdsdev** [comment]
@dsyrstad the `Prefer: params=single-object` has been removed from PostgREST https://github.com/PostgREST/postgrest/pull/3757

**Comment by dsyrstad** [comment]
> [@dsyrstad](https://github.com/dsyrstad) the `Prefer: params=single-object` has been removed from PostgREST [PostgREST/postgrest#3757](https://github.com/PostgREST/postgrest/pull/3757)

Yes, it looks like the `Prefer: params=single-object header` header is gone now in v14, but we also use `Prefer: tx=rollback` and `Prefer: count=exact`, which are still supported in v14 (https://docs.postgrest.org/en/stable/references/api/preferences.html#prefer-header). There is no way to set these headers with the Dart `rpc()` API.

BTW, we use `Prefer: count=exact` as a hacky workaround to https://github.com/PostgREST/postgrest/discussions/2163.

Here's our extension that provides an `rpcWith()` method to workaround the header problems:

```dart
extension SupabaseClientExtension on SupabaseClient {
  /// Calls an rpc with the given header parameters.
  Future<dynamic> rpcWith(
    String fn, {
    Map<String, String>? customHeaders,
    bool? singleObject,
    String? txOption,
    Map<String, dynamic>? params,
  }) async {
    final newHeaders = <String, String>{
      ...?customHeaders,
      'Prefer': [
        if (singleObject == true) 'params=single-object',
        if (txOption != null) 'tx=$txOption',
        // NOTE: This causes PostgREST NOT to try to re-execute the function for a total count.
        // It is a hacky workaround. See https://github.com/PostgREST/postgrest/discussions/2163.
        'count=exact',
      ].join(','),
    };

    final localRest = rest.schema('public');
    localRest.headers.addAll({...rest.headers, ...headers, ...newHeaders});

    final response = await localRest.rpc(fn, params: params);
    // The Prefer: count=exact causes it to return a PostgrestResponse object rather
    // the raw data - sometimes.
    if (response is PostgrestResponse) {
      return response.data;
    }

    return response;
  }
}
```

This implementation was based on v12, so the `singleObject` param/logic could be removed now.

**Comment by Vinzent03** [comment]
@dsyrstad I think what you are looking for is `supabase.rpc("functionname").count(CountOption.exact)`, which is already available.

**Comment by dsyrstad** [comment]
> [@dsyrstad](https://github.com/dsyrstad) I think what you are looking for is `supabase.rpc("functionname").count(CountOption.exact)`, which is already available.

@Vinzent03  Thank you for that! I was not aware that you could do that. We also need the `Prefer: tx=...` header, but looking into how `rpc().count()` works internally, I see there is now a `setHeader()` function in `PostgrestFilterBuilder`, so we can use that to set the headers.  

@grdsdev  I think that `PostgrestFilterBuilder.setHeaders()` solves this issue completely, and it can remain closed.
